modules-magic-drc/run_all_drc.py

```diff
@@ -0,0 +1,150 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+#-------------------------------------------------------------------------
+# run_all_drc.py --- A script that will run run_standard_drc
+# for all .gds files under the cells/ folder. Must be run from repository
+# root. Author: Mohamed Gaber
+# 
+# Usage:
+#
+#   run_all_drc.py [optional subdirectory filter]
+#
+# Results:
+#
+#   Prints a report to standard output.
+# 	
+#-------------------------------------------------------------------------
+
+import os
+import re
+import sys
+import click
+import traceback
+import subprocess
+import concurrent.futures as fut
+
+ACCEPTABLE_ERRORS = [
+    "All nwells must contain metal-connected N+ taps (nwell.4)",
+    "P-diff distance to N-tap must be < 15.0um (LU.3)",
+    "N-diff distance to P-tap must be < 15.0um (LU.2)",
+]
+
+SCRIPT_DIR = os.path.realpath(os.path.dirname(__file__))
+STANDARD_DRC_SCRIPT = os.path.join(SCRIPT_DIR, "run_standard_drc.py")
+PDK_SUBSET = os.path.join(SCRIPT_DIR, "pdk_subset")
+
+def parse_drc_report(report):    
+    components = list(map(lambda x: x.split("\n"), report.split("\n\n")))
+    
+    header = components[0]
+    cell_name = header[0][28:]
+
+    errors = []
+    for error in components[1:]:
+        error_name = error[0]
+        if error_name in ACCEPTABLE_ERRORS:
+            continue
+        errors.append(error)
+    
+    return errors
+
+def drc_gds(path):
+    cell_name = os.path.basename(path)[:-4]
+    
+    env = os.environ.copy()
+    env["PDKPATH"] = PDK_SUBSET
+
+    res = subprocess.run([
+        "python3",
+        STANDARD_DRC_SCRIPT,
+        path
+    ], env=env, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    
+    report_path = path[:-4] + "_drc.txt"
+    try:
+        report = open(report_path).read()
+
+        return cell_name, parse_drc_report(report)
+    except FileNotFoundError:
+        return cell_name, [["Magic did not produce a report.", res.stdout.decode("utf8"), res.stderr.decode("utf8")]]
+
+@click.command()
+@click.option("-m", "--match-directories", default=".", help="A regex that will match subdirectories under cells/. Default: . (matches everything.)")
+@click.option("-b", "--known-bad", default="", help="A comma,delimited list of cells that are known bad and thus do not cause a non-zero exit upon failure. Default: empty string (None of them.)")
+def run_all_drc(match_directories, known_bad):
+    print("Testing cells in directories matching /%s/…" % match_directories)
+
+    known_bad_list = known_bad.split(",")
+
+    nproc = os.cpu_count()
+    with fut.ThreadPoolExecutor(max_workers=nproc) as executor:
+        futures = []
+
+        cells_dir = "./cells"
+        cells = os.listdir(cells_dir)
+
+        print("Submitting futures…")
```

Comment:
This seem like debugging messages? What does this output mean for a user?

Reply:
Yeah, sorry about that.